src/messenger.rs

```diff
@@ -922,6 +932,122 @@ mod tests {
         assert_eq!(actual, resp);
     }
 
+    #[tokio::test]
+    async fn test_cancel_request() {
+        // Use a "virtual" network between a simulated broker and a client. The network is intercepted in the middle to
+        // pause it after 3 bytes are sent by the client.
+        let (tx_front, rx_middle) = tokio::io::duplex(1);
+        let (tx_middle, mut rx_back) = tokio::io::duplex(1);
+
+        let messenger = Messenger::new(tx_front, 1_000);
+
+        let network_pause = Arc::new(Barrier::new(2));
+        let network_pause_captured = Arc::clone(&network_pause);
+        let handle_network = tokio::spawn(async move {
+            // Need to split both directions into read and write halfs so we can run full-deplux in two loops. Otherwise
+            // the test might deadlock even though the code is just fine (TCP is full-duplex).
+            let (mut rx_middle_read, mut rx_middle_write) = tokio::io::split(rx_middle);
+            let (mut tx_middle_read, mut tx_middle_write) = tokio::io::split(tx_middle);
+
+            let direction1 = async {
+                for i in 0.. {
+                    let mut buf = [0; 1];
+                    rx_middle_read.read_exact(&mut buf).await.unwrap();
+                    tx_middle_write.write_all(&buf).await.unwrap();
+
+                    if i == 3 {
+                        network_pause_captured.wait().await;
+                    }
+                }
+            };
+
+            let direction2 = async {
+                loop {
+                    let mut buf = [0; 1];
+                    tx_middle_read.read_exact(&mut buf).await.unwrap();
+                    rx_middle_write.write_all(&buf).await.unwrap();
+                }
+            };
+
+            tokio::select! {
+                _ = direction1 => {}
+                _ = direction2 => {}
+            }
+        });
+
+        // simulated broker, just reads messages and answers w/ "api versions" responses
+        let handle_broker = tokio::spawn(async move {
+            for correlation_id in 0.. {
+                rx_back.read_message(1_000).await.unwrap();
```

Comment:
Could possibly check that the message is an ApiVersions request

Reply:
done